### PR TITLE
improvement(ldap authorization): adding support to LDAP

### PR DESCRIPTION
**Pull request by fgelcer**
only authorization right now.
it is the whole code needed, but still no tests
were yet set to run with LDAP, but there is now
a configurations yaml to add to any test to run
with the LDAP.

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [x] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: sdcm/cluster.py
Comment (fruch) [inline sdcm/cluster.py:2848]: please remove the `default` param, None is the default if flag wasn't defined

File: sdcm/cluster.py
Comment (fruch) [inline sdcm/cluster.py:125]: why do need all those on the Setup class ? they are going to change during the test ?

File: sdcm/localhost.py
Comment (fruch) [inline sdcm/localhost.py:44]: how this configuration is related to the localhost class ? i'm not sure I'm following...

File: sdcm/cluster.py
Comment (fruch) [inline sdcm/cluster.py:2744]: maybe `ldap_docker_running` ?

**Comment by roydahan** [comment]
@scylladb/qa-maintainers please note that we have now 2 PRs that change the hydra version.
The first one should win (this) and @juliayakovlev will need to include this one on top of her change.

**Comment by fruch** [comment]
> @scylladb/qa-maintainers please note that we have now 2 PRs that change the hydra version.
> The first one should win (not sure which exactly) and maybe @juliayakovlev will need to include this one on top of her change or the opossite.

the first to be merged should win :smile: 
the looser get to rebuild his docker image after he rebase

File: sdcm/cluster.py
Comment (fgelcer) [reply sdcm/cluster.py:125]: they will be populated once the container is created, and will be used across the test in different places...
the `LDAP_ADDRESS` will be used for create a connection to the LDAP server for example...

File: sdcm/localhost.py
Comment (fgelcer) [reply sdcm/localhost.py:44]: the localhost configuration runs on the runner, and these configurations (or most of them) are sent to the container creation parameters...

File: sdcm/cluster.py
Comment (fgelcer) [reply sdcm/cluster.py:2848]: ok

File: sdcm/cluster.py
Comment (fgelcer) [reply sdcm/cluster.py:2744]: ok

File: sdcm/cluster.py
Comment (fruch) [reply sdcm/cluster.py:125]: the `LDAP_ADDRESS` o.k. but the rest are hardcodede values for now, like username/password and such. I don't see the reason for them to be on Setup class.

in some places you are using `LDAP_BASE_OBJECT` directly, and in some reading it from ` node.ldap_config['ldap_base_object']` and in some from `Setup.LDAP_BASE_OBJECT`

I would prefer it would be the first one only, until prove it's need to be dynamic or change during the run.
i.e. all those are known in advance, no need to put them all over the place.

**Comment by fruch** [comment]
please rebase, and update the version in `docker/env/version` again and build the docker again.
next time please give a temporary name when building a new docker, i.e. `0.78_add_ldap`, and when getting close to approval, can be given the actual version and rebuild it.

**Comment by fruch** [comment]
> > @scylladb/qa-maintainers please note that we have now 2 PRs that change the hydra version.
> > The first one should win (not sure which exactly) and maybe @juliayakovlev will need to include this one on top of her change or the opossite.
> 
> the first to be merged should win 
> the looser get to rebuild his docker image after he rebase

@fgelcer BTW you lost to @juliayakovlev  :stuck_out_tongue:

File: sdcm/cluster.py
Comment (fgelcer) [reply sdcm/cluster.py:125]: the only access to `node.ldap_config['ldap_base_object']` is when we set the `Setup.LDAP_BASE_OBJECT`, so its usage is pretty consistent

**Comment by fgelcer** [comment]
> > > @scylladb/qa-maintainers please note that we have now 2 PRs that change the hydra version.
> > > The first one should win (not sure which exactly) and maybe @juliayakovlev will need to include this one on top of her change or the opossite.
> > 
> > 
> > the first to be merged should win
> > the looser get to rebuild his docker image after he rebase
> 
> @fgelcer BTW you lost to @juliayakovlev

it was intentional, as ladies first :-)

File: docker/env/version
Comment (bentsi) [inline docker/env/version:1]: You broke Hydra version 0.80, this is not the way we update the Hydra Docker image. Did you talk with somebody before doing this?

File: sdcm/cluster.py
Comment (fruch) [reply sdcm/cluster.py:125]: you are missing the point, having one `LDAP_BASE_OBJECT` that all the code point to it, if it's a const value.
no need for all the wrapper around it.

`Setup` should be used for global values we are changing during he span of test, i.e. like addresses and ports.

Please remove any consts values out of `Setup` class.

File: docker/env/version
Comment (fruch) [reply docker/env/version:1]: also version `0.79` was burnt...
anyhow next time I think it would be clear, when building a version docker for testing a PR,
it should be `0.80_[some suffix to describe it]`, it should be rebuild into a formal version only once the PR is approved.
(which isn't the case yet)

**Review by fruch — CHANGES_REQUESTED** [review]
static vaule should go out of `Setup` class
* LDAP_USERNAME
* LDAP_PASSWORD
* LDAP_BASE_OBJECT
* LDAP_USERS
* LDAP_ROLE

File: docker/env/version
Comment (fgelcer) [reply docker/env/version:1]: > You broke Hydra version 0.80, this is not the way we update the Hydra Docker image. Did you talk with somebody before doing this?

what was broken with v0.80?

yes, I spoke the guys, and as @fruch  mentioned in his comment [here](https://github.com/scylladb/scylla-cluster-tests/pull/2761#issuecomment-704538833), next time, i will create it once getting close to approval (as I expected it to be) or once approved (and then change the version file + build and push the image)...

**Comment by fgelcer** [comment]
> static vaule should go out of `Setup` class
> 
> * LDAP_USERNAME
> * LDAP_PASSWORD
> * LDAP_BASE_OBJECT
> * LDAP_USERS
> * LDAP_ROLE

why is the `LDAP_ADDRESS` different from these other parameters?

**Comment by fruch** [comment]
@fgelcer this would need rebuilding of the docker cause of #2789

**Comment by roydahan** [comment]
@bentsi / @fruch Except of the docker version issue that @fgelcer will handle right before merging (so it won't rotten).
Any other feedback?
If not, please mark as approved.

File: sdcm/localhost.py
Comment (fruch) [inline sdcm/localhost.py:44]: this property can be remove, it's not used anywhere

**Review by fruch — CHANGES_REQUESTED** [review]
`localhost.ldap_config` property can be removed, since it's not used. (and need to fix a conflict on this file anyhow)
beside that, looks o.k.

**Comment by fgelcer** [comment]
> `localhost.ldap_config` property can be removed, since it's not used. (and need to fix a conflict on this file anyhow)
> beside that, looks o.k.

everything was updated...
just, once approved, and before it will be merged, the `docker/env/version` need to be updated to the final value of it

**Comment by fgelcer** [comment]
conflicts were resolved, and tests are making very good progress.
ping @fruch @roydahan

File: configurations/ldap-authorization.yaml
Comment (roydahan) [inline configurations/ldap-authorization.yaml:4]: Is this what required when using LDAP?

File: configurations/ldap-authorization.yaml
Comment (roydahan) [inline configurations/ldap-authorization.yaml:3]: Why are these relevant to this configuration?

File: sdcm/cluster_aws.py
Comment (roydahan) [inline sdcm/cluster_aws.py:870]: the default should be in sct_config

File: sdcm/cluster_k8s/__init__.py
Comment (roydahan) [inline sdcm/cluster_k8s/__init__.py:587]: same

**Review by roydahan — COMMENTED** [review]
Overall looks good to me.
Had some minor comments you need to address.

In general, you're using too much positional argument when calling functions, which is bad practice for future additions and refactors.
I prefer you would use keyword arguments like: ip=ip, port=port, etc.

File: configurations/ldap-authorization.yaml
Comment (fgelcer) [reply configurations/ldap-authorization.yaml:4]: yes, as this is the "authorizer"

File: configurations/ldap-authorization.yaml
Comment (fgelcer) [reply configurations/ldap-authorization.yaml:3]: as we are setting the `authorizer`, and if the user and password weren't set up before, then it need to be set up in here altogether, otherwise, it won't work

Comment (roydahan) [reply configurations/ldap-authorization.yaml:3]: But this is the authenticator, not autorizer.

Comment (fgelcer) [reply configurations/ldap-authorization.yaml:3]: this is what will allow one user to enter a username and password

File: sdcm/cluster_aws.py
Comment (fgelcer) [reply sdcm/cluster_aws.py:870]: how do you set this on sct_config?

File: sdcm/cluster_k8s/__init__.py
Comment (fgelcer) [reply sdcm/cluster_k8s/__init__.py]: removed

**Comment by fgelcer** [comment]
@roydahan / @fruch ping

**Comment by fgelcer** [comment]
> @roydahan / @fruch ping

rebased on latest master, and already resolved the conflicts there were with this PR

**Comment by fgelcer** [comment]
@fruch @roydahan , i see that someone created a new `docker/env/version` so once this one is approved (and ready to merge) i will re-create another docker image and push it to dockerhub

**Comment by fgelcer** [comment]
@roydahan , fixed conflicts (changed the docker version) and added the 2 pipelines we spoke of. I will now, while this isn't yet approved and merged, create the jobs (for enterprise and 2020.1) and will give them a ride.
@fruch , ready to merge (as you have already reviewed it)

**Review by roydahan — APPROVED** [review]

**Comment by fgelcer** [comment]
@fruch / @roydahan , i have created the jobs for 2020.1 and enterprise, and i ran the 2020.1 job (short longevity with LDAP enabled) and it passed:
https://jenkins.scylladb.com/view/enterprise-2020.1/job/enterprise-2020.1/job/SCT_Enterprise_Features/job/LDAP_authorization/job/longevity-100gb-4h-with-ldap-authorization/

**Comment by fgelcer** [comment]
@fruch ping

**Review by fruch — APPROVED** [review]
LGTM

**Comment by fruch** [comment]
Fabio, please upload the formal docker, so it can be merged

**Comment by fgelcer** [comment]
> Fabio, please upload the formal docker, so it can be merged

@fruch , done

**Comment by fgelcer** [comment]
@scylladb/qa-maintainers ping

**Review by fruch — APPROVED** [review]
LGTM

**Pull request merged**
